kernel/comps/network/src/dma_pool.rs

```diff
@@ -0,0 +1,254 @@
+#[derive(Debug)]
+pub struct DmaPool {
+    direction: DmaDirection,
+    block_size: usize,
+    is_cache_coherent: bool,
+    pages: RwLock<VecDeque<Arc<DmaPage>>>,
+}
+
+#[derive(Debug)]
+struct DmaPage {
+    storage: DmaStream,
+    // `BitArray` is 64 bits, since each `DmaBlock` is bigger than 64Bytes, there's no more
+    // than `PAGE_SIZE` / 64 = 64 `DmaBlock`s in a `DmaPage`.
+    allocated_blocks: SpinLock<BitArray>,
+}
+
+impl DmaPage {
+    fn alloc_block(self: &Arc<Self>, block_size: usize) -> Option<DmaBlock> {
+        let mut blocks = self.allocated_blocks.lock_irq_disabled();
+
+        let free_block_index = blocks.iter_zeros().nth(0)?;
+        if free_block_index * block_size >= self.storage.nbytes() {
+            return None;
+        }
+        blocks.set(free_block_index, true);
+
+        let block = DmaBlock {
+            dma_stream: self.storage.clone(),
+            start_addr: self.storage.daddr() + free_block_index * block_size,
+            size: block_size,
+            page: Arc::downgrade(self),
+        };
+
+        Some(block)
+    }
+
+    fn is_free(&self) -> bool {
+        *self.allocated_blocks.lock() == BitArray::<[usize; 1], Lsb0>::ZERO
+    }
+}
+
+impl HasDaddr for DmaPage {
+    fn daddr(&self) -> Daddr {
+        self.storage.daddr()
+    }
+}
+
+/// Small streaming DMA areas. The size of `DmaBlock` ranges from 64 bytes to `PAGE_SIZE` and must be 2^K.
+#[derive(Debug)]
+pub struct DmaBlock {
+    dma_stream: DmaStream,
+    start_addr: Daddr,
+    size: usize,
+    page: Weak<DmaPage>,
+}
+
+impl HasDaddr for DmaBlock {
+    fn daddr(&self) -> Daddr {
+        self.start_addr
+    }
+}
+
+impl DmaBlock {
+    pub const fn size(&self) -> usize {
+        self.size
+    }
+
+    pub fn reader(&self) -> Result<VmReader<'_>, aster_frame::Error> {
+        let offset = self.start_addr - self.dma_stream.daddr();
+        Ok(self.dma_stream.reader()?.skip(offset).limit(self.size))
+    }
+
+    pub fn writer(&self) -> Result<VmWriter<'_>, aster_frame::Error> {
+        let offset = self.start_addr - self.dma_stream.daddr();
+        Ok(self.dma_stream.writer()?.skip(offset).limit(self.size))
+    }
+
+    pub fn sync(&self, byte_range: Range<usize>) -> Result<(), aster_frame::Error> {
+        let offset = self.daddr() - self.dma_stream.daddr();
+        let range = byte_range.start + offset..byte_range.end + offset;
+        self.dma_stream.sync(range)
+    }
+}
+
+impl Drop for DmaBlock {
+    fn drop(&mut self) {
+        let page = self.page.upgrade().unwrap();
+        debug_assert!((page.daddr()..page.daddr() + PAGE_SIZE).contains(&self.daddr()));
+        let block_idx = (self.daddr() - page.daddr()) / self.size;
+        page.allocated_blocks.lock().set(block_idx, false);
+    }
+}
+
+impl DmaPool {
+    pub fn new(block_size: usize, direction: DmaDirection, is_cache_coherent: bool) -> Self {
```

Comment:
I think the `block_size` can be a `const genetic`. This is because the size of `DmaBlock` is actually determined at the time when `DmaPool` is created, and this size will not change afterwards.